File: Java/src/main/java/com/DSA/Java/Tree/BinaryHeap/BinaryHeap.java
package com.DSA.Java.Tree.BinaryHeap;

public class BinaryHeap {
   int binaryHeapArray[];
   int sizeOfArray;

   public BinaryHeap(int size) {
      this.binaryHeapArray = new int[size + 1];
      sizeOfArray = 0;
   }
   
   public boolean isEmpty() {
      return sizeOfArray == 0;
   }
   
   public boolean isFull() {
      return sizeOfArray == binaryHeapArray.length - 1;
   }

   public Integer peek() {
      if (isEmpty()) {
         System.out.println("The Heap is empty");
         return null;
      }
      return binaryHeapArray[1];
   }

   public int sizeOfHeap() {
      return sizeOfArray;
   }

   public void levelOrderTraversal() {
      for (int i = 1; i <= sizeOfArray; i++) {
         System.out.print(binaryHeapArray[i] + " ");
      }
      System.out.println();
   }

   public void preOrderTraversal(int index) {
      if (index > sizeOfArray) {
         return ;
      }
      System.out.print(binaryHeapArray[index] + " ");
      preOrderTraversal(index * 2);
      preOrderTraversal(index * 2 + 1);
   }

   public void inOrderTraversal(int index) {
      if (index > sizeOfArray) {
         return ;
      }
      inOrderTraversal(index * 2);
      System.out.print(binaryHeapArray[index] + " ");
      inOrderTraversal(index * 2 + 1);
   }

   public void postOrderTraversal(int index) {
      if (index > sizeOfArray) {
         return ;
      }
      postOrderTraversal(index * 2);
      postOrderTraversal(index * 2 + 1);
      System.out.print(binaryHeapArray[index] + " ");
   }
   
   private void heapifyBottomToTop(int index, String heapType) {
      if (index <= 1) {
         return;
      }
      if (heapType == "Max") {
         if (binaryHeapArray[index] > binaryHeapArray[index / 2]) {
            int temp = binaryHeapArray[index];
            binaryHeapArray[index] = binaryHeapArray[index / 2];
            binaryHeapArray[index / 2] = temp;
         }  
      } else if (heapType == "Min") {
         if (binaryHeapArray[index] < binaryHeapArray[index / 2]) {
            int temp = binaryHeapArray[index];
            binaryHeapArray[index] = binaryHeapArray[index / 2];
            binaryHeapArray[index / 2] = temp;
         }
      }
      heapifyBottomToTop(index / 2, heapType);
   }

   public void insertElement(int value, String heapType) {
      if (isFull()) {
         System.out.println("The Heap is already full");
         return;
      }
      binaryHeapArray[sizeOfArray + 1] = value;
      sizeOfArray++;
      heapifyBottomToTop(sizeOfArray, heapType);
   }
   
   private void heapifyTopToBottom(int index, String heapType) {
      int left = index * 2;
      int right = index * 2 + 1;
      int swapped = 0;

      if (index > sizeOfArray) {
         return;
      }
      if (heapType == "Min") {
         if (sizeOfArray == left && binaryHeapArray[index] < binaryHeapArray[left]) {
            int temp = binaryHeapArray[index];
            binaryHeapArray[index] = binaryHeapArray[left];
            binaryHeapArray[left] = temp;
            return ;
         } else {
            if (binaryHeapArray[left] < binaryHeapArray[right]) {
               swapped = left;
            } else {
               swapped = right;
            }

            if (binaryHeapArray[index] > binaryHeapArray[swapped]) {
               int temp = binaryHeapArray[index];
               binaryHeapArray[index] = binaryHeapArray[swapped];
               binaryHeapArray[swapped] = temp;
            }
         }
      } else if (heapType == "Max") {
         if (sizeOfArray == right && binaryHeapArray[index] > binaryHeapArray[right]) {
            int temp = binaryHeapArray[index];
            binaryHeapArray[index] = binaryHeapArray[right];
            binaryHeapArray[right] = temp;
            return ;
         } else {
            if (binaryHeapArray[left] > binaryHeapArray[right]) {
               swapped = left;
            } else {
               swapped = right;
            }

            if (binaryHeapArray[index] < binaryHeapArray[swapped]) {
               int temp = binaryHeapArray[index];
               binaryHeapArray[index] = binaryHeapArray[swapped];
               binaryHeapArray[swapped] = temp;
            }
         }
      }
      heapifyTopToBottom(swapped, heapType);
   }

   public int extractTopElement(String heapType) {
      if (isEmpty()) {
         return -1;
      }
      int extractedValue = binaryHeapArray[1];
      binaryHeapArray[1] = binaryHeapArray[sizeOfArray];
      sizeOfArray--;
      heapifyTopToBottom(1, heapType);

      return extractedValue;
   }
}

File: C++/Tree/src/BinaryHeap/BinaryHeap.cpp
#include "../../include/BinaryHeap/BinaryHeap.h"

BinaryHeap::BinaryHeap(int size, const HeapType type) {
    this->heap = new int[size + 1];
    this->size_of_heap = 0;
    this->size_of_array = size + 1;
    this->type_of_heap = type;
}

bool BinaryHeap::is_empty() const {
    return size_of_heap == 0;
}

bool BinaryHeap::is_full() const {
    return size_of_heap >= size_of_array - 1;
}

int BinaryHeap::get_size_of_heap() const {
    return size_of_heap;
}

int BinaryHeap::peek() const {
    if (is_empty()) {
        std::cout << "The Heap is empty" << std::endl;
        return -1;
    }
    return heap[1];
}

void BinaryHeap::level_order_traversal() {
    if (is_empty()) {
        std::cout << "The Heap is empty" << std::endl;
        return ;
    }

    for (int i = 1; i <= get_size_of_heap(); i++) {
        std::cout << heap[i] << " ";
    }
    std::cout << "\n";
}

void BinaryHeap::in_order_traversal(int index) {
    if (index > get_size_of_heap()) {
        return ;
    }
    std::cout << heap[index] << " ";
    in_order_traversal(index * 2);
    in_order_traversal(index * 2 + 1);
}

void BinaryHeap::pre_order_traversal(int index) {
    if (index > get_size_of_heap()) {
        return ;
    }
    pre_order_traversal(index * 2);
    std::cout << heap[index] << " ";
    pre_order_traversal(index * 2 + 1);
}

void BinaryHeap::post_order_traversal(int index) {
    if (index > get_size_of_heap()) {
        return ;
    }
    post_order_traversal(index * 2);
    post_order_traversal(index * 2 + 1);
    std::cout << heap[index] << " ";
}

void heapify_bottom_to_top(int index, const std::string &type, int *heap) {
    if (index <= 1) {
        return ;
    }
    if (type == "max") {
        if (heap[index] > heap[index / 2]) {
            int temp = heap[index];
            heap[index] = heap[index / 2];
            heap[index / 2] = temp;
        }
    } else if (type == "min") {
        if (heap[index] < heap[index / 2]) {
            int temp = heap[index];
            heap[index] = heap[index / 2];
            heap[index / 2] = temp;
        }
    }
    heapify_bottom_to_top(index / 2, type, heap);
}

void BinaryHeap::insert_element(int value) {
    if (is_full()) {
        std::cout << "The Heap is full" << std::endl;
        return ;
    }
    size_of_heap++;
    heap[size_of_heap] = value;

    std::string type_of_oper = type_of_heap == HeapType::MAX ? "max" : "min";

    heapify_bottom_to_top(size_of_heap, type_of_oper, heap);
}

void heapify_top_to_bottom(int index, const std::string &type, int *heap, int size_of_heap) {
    int left {index * 2}, right {index * 2 + 1}, swapped {0};
    if (left > size_of_heap) {
        return ;
    }

    if (type == "min") {
        if (size_of_heap == left && heap[index] < heap[left]) {
            int temp = heap[index];
            heap[index] = heap[left];
            heap[left] = temp;
            return ;
        } else {
            if (heap[left] < heap[right]) {
                swapped = left;
            } else {
                swapped = right;
            }

            if (heap[index] > heap[swapped]) {
                int temp = heap[index];
                heap[index] = heap[swapped];
                heap[swapped] = temp;
            }
        }
    } else if (type == "max") {
        if (size_of_heap == right && heap[index] > heap[right]) {
            int temp = heap[index];
            heap[index] = heap[right];
            heap[right] = temp;
        } else {
            if (heap[left] > heap[right]) {
                swapped = left;
            } else {
                swapped = right;
            }

            if (heap[index] < heap[swapped]) {
                int temp = heap[index];
                heap[index] = heap[swapped];
                heap[swapped] = temp;
            }
        }
    }
    heapify_top_to_bottom(swapped, type, heap, size_of_heap);
}

int BinaryHeap::extract_top_element() {
    int extracted_value = heap[1];
    heap[1] = heap[get_size_of_heap()];
    size_of_heap--;

    std::string type_of_oper = type_of_heap == HeapType::MAX ? "max" : "min";

    heapify_top_to_bottom(1, type_of_oper, heap, get_size_of_heap() - 1);

    return extracted_value;
}

void BinaryHeap::delete_heap() {
    heap = nullptr;
    size_of_heap = 0;
    size_of_array = 0;
}

BinaryHeap::~BinaryHeap() {
    delete [] heap;
}
